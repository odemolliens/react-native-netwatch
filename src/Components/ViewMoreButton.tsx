import * as React from 'react';
import { useContext } from 'react';
import { TouchableOpacity, Text, StyleSheet } from 'react-native';
import { ThemeContext } from '../Theme';
import MaterialCommunityIcons from 'react-native-vector-icons/MaterialCommunityIcons';

export interface IProps {
  onPress: () => void;
}

// Button use in the details screens to go in the JSON tree view. It's just a button text.
export const ViewMoreButton: React.FC<IProps> = (props: IProps) => {
  const theme = useContext(ThemeContext);

  return (
    <TouchableOpacity
      testID="buttonViewMore"
      style={styles.container}
      onPress={() => {
        props.onPress();
      }}
    >
      <Text style={[{ color: theme.primaryColor, marginRight: 6 }]}>View more</Text>
      <MaterialCommunityIcons
        name="more"
        color={theme.primaryColor}
        size={14}
        style={{ transform: [{ rotateY: '180deg' }] }}
      />
    </TouchableOpacity>
  );
};

export default ViewMoreButton;

const styles = StyleSheet.create({
  container: {
    flexDirection: 'row',
    borderLeftWidth: 0,
    alignItems: 'center',
  },
});
