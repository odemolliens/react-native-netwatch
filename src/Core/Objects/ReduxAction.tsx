import { ILog, LogType } from '../../types';

interface IAction {
  type: string;
  payload: any;
}

export class ReduxAction implements ILog {
  _id: number = -1;
  type: LogType = 'REDUX';
  startTime: number = Date.now();
  stringifiedAction: string = '';
  extra: string = '';
  action: IAction = { type: '__ERROR:UNDEFINED__', payload: '' };

  constructor(attributes?: any) {
    Object.assign(this, attributes);
  }
}

export default ReduxAction;
