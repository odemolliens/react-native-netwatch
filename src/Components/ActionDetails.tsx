import * as React from 'react';
import { useContext } from 'react';
import { View, StyleSheet } from 'react-native';
import { Subheading } from 'react-native-paper';
import { reduxTag } from './Status';
import { getTime, getShortDate, isLongText, addEllipsis } from '../Utils/helpers';
import { Text, Title } from './Text';
import { ThemeContext } from '../Theme';
import ReduxAction from '../Core/Objects/ReduxAction';
import { ViewMoreButton } from './ViewMoreButton';

export interface IProps {
  testId?: string;
  item: ReduxAction;
  reduxAction: any;
  onPressViewMore: (value: boolean) => void;
}

export const ActionDetails: React.FC<IProps> = (props: IProps) => {
  const theme = useContext(ThemeContext);

  return (
    <View style={{ flex: 1, width: '100%' }}>
      {/* REDUX + REDUX TAG */}
      <View style={[styles.line]}>
        <Title style={[{ marginRight: 6 }, { color: theme.textColorOne }]}>REDUX</Title>
        {reduxTag()}
      </View>

      <View style={[styles.line]}>
        <Text style={{ color: theme.textColorFour }}>Started at : </Text>
        <Text>{`${getShortDate(props.item.startTime)} - ${getTime(props.item.startTime)}`}</Text>
      </View>

      <Subheading
        style={[styles.subheading, { backgroundColor: theme.secondaryLightColor, color: theme.textColorOne }]}
      >
        Content
      </Subheading>

      <View style={[styles.line]}>
        <Text style={{ color: theme.textColorFour }}>Type</Text>
      </View>
      <View style={[styles.line]}>
        <Text>{props.item.action.type}</Text>
      </View>

      <View style={[styles.line]}>
        {props.reduxAction.label && (
          <Text style={{ color: theme.textColorFour }}>{`${props.reduxAction.label
            .charAt(0)
            .toUpperCase()}${props.reduxAction.label.slice(1)} :`}</Text>
        )}
      </View>
      <View style={[styles.line]}>
        <View style={[styles.attribtuesContainer, { flex: 1, paddingHorizontal: 0 }]}>
          {props.reduxAction.payload && <Text>{addEllipsis(JSON.stringify(props.reduxAction.payload, null, 2))}</Text>}
          {isLongText(addEllipsis(JSON.stringify(props.reduxAction.payload, null, 2))) && (
            <View style={{ alignItems: 'flex-end' }}>
              <ViewMoreButton onPress={() => props.onPressViewMore(true)} />
            </View>
          )}
        </View>
      </View>
    </View>
  );
};

export default ActionDetails;

const styles = StyleSheet.create({
  line: {
    flexDirection: 'row',
    paddingHorizontal: 16,
    marginBottom: 10,
    alignItems: 'center',
  },

  subheading: {
    paddingHorizontal: 16,
    paddingVertical: 8,
    marginBottom: 10,
  },

  attribtuesContainer: {
    justifyContent: 'flex-start',
    paddingHorizontal: 16,
    paddingVertical: 8,
  },
});
