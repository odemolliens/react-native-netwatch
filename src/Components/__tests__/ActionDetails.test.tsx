import * as React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import { ActionDetails, IProps } from '../ActionDetails';
import ReduxAction from '../../Core/Objects/ReduxAction';
import ViewMoreButton from '../ViewMoreButton';

describe('Status test suite', () => {
  let component: ShallowWrapper;
  let props: IProps;
  const onPressViewMore = jest.fn();

  it('should render properly redux action', () => {
    givenProps(mockAction, {
      label: '__ERROR:UNDEFINED__',
      payload:
        'This is a very long payload to be sure that we have more than 100 characters to explore all the branches present in this component',
    });
    givenComponent();
    expect(component).toMatchSnapshot();
  });

  it('should execute props.onPressViewMore', () => {
    givenProps(mockAction, {
      label: '__ERROR:UNDEFINED__',
      payload:
        'This is a very long payload to be sure that we have more than 100 characters to explore all the branches present in this component',
    });
    givenComponent();
    component.find(ViewMoreButton).simulate('press');
    expect(component).toMatchSnapshot();
  });

  // GIVEN
  function givenComponent() {
    component = shallow(<ActionDetails {...props} />);
  }

  function givenProps(item: ReduxAction, reduxAction) {
    props = {
      item,
      reduxAction,
      onPressViewMore,
    };
  }

  const mockAction: ReduxAction = new ReduxAction({
    _id: 73,
    startTime: 100,
    type: 'REDUX',
    stringifiedAction: "{ type: '__ERROR:UNDEFINED__', action: '' }",
    action: { type: '__ERROR:UNDEFINED__', action: '' },
  });
});
