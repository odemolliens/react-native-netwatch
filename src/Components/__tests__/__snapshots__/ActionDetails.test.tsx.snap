// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Status test suite should execute props.onPressViewMore 1`] = `
<Component
  style={
    Object {
      "flex": 1,
      "width": "100%",
    }
  }
>
  <Component
    style={
      Array [
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "marginBottom": 10,
          "paddingHorizontal": 16,
        },
      ]
    }
  >
    <Title
      style={
        Array [
          Object {
            "marginRight": 6,
          },
          Object {
            "color": "#F9FAFB",
          },
        ]
      }
    >
      REDUX
    </Title>
    <Component
      style={
        Array [
          Object {
            "borderRadius": 4,
            "fontSize": 12,
            "height": 22,
            "justifyContent": "center",
            "width": 30,
          },
          Object {
            "backgroundColor": "#764ABC",
          },
          Object {
            "paddingHorizontal": 8,
          },
        ]
      }
      testID="statusCodeColor"
    >
      <Icon
        color="#F9FAFB"
        name="redux"
        size={12}
      />
    </Component>
  </Component>
  <Component
    style={
      Array [
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "marginBottom": 10,
          "paddingHorizontal": 16,
        },
      ]
    }
  >
    <Text
      style={
        Object {
          "color": "#6B7280",
        }
      }
    >
      Started at : 
    </Text>
    <Text>
      01/01/1970 - 00:00:00
    </Text>
  </Component>
  <Subheading
    style={
      Array [
        Object {
          "marginBottom": 10,
          "paddingHorizontal": 16,
          "paddingVertical": 8,
        },
        Object {
          "backgroundColor": "#374151",
          "color": "#F9FAFB",
        },
      ]
    }
  >
    Content
  </Subheading>
  <Component
    style={
      Array [
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "marginBottom": 10,
          "paddingHorizontal": 16,
        },
      ]
    }
  >
    <Text
      style={
        Object {
          "color": "#6B7280",
        }
      }
    >
      Type
    </Text>
  </Component>
  <Component
    style={
      Array [
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "marginBottom": 10,
          "paddingHorizontal": 16,
        },
      ]
    }
  >
    <Text>
      __ERROR:UNDEFINED__
    </Text>
  </Component>
  <Component
    style={
      Array [
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "marginBottom": 10,
          "paddingHorizontal": 16,
        },
      ]
    }
  >
    <Text
      style={
        Object {
          "color": "#6B7280",
        }
      }
    >
      __ERROR:UNDEFINED__ :
    </Text>
  </Component>
  <Component
    style={
      Array [
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "marginBottom": 10,
          "paddingHorizontal": 16,
        },
      ]
    }
  >
    <Component
      style={
        Array [
          Object {
            "justifyContent": "flex-start",
            "paddingHorizontal": 16,
            "paddingVertical": 8,
          },
          Object {
            "flex": 1,
            "paddingHorizontal": 0,
          },
        ]
      }
    >
      <Text>
        "This is a very long payload to be sure that we have more than 100 characters to explore all the bra...
      </Text>
      <Component
        style={
          Object {
            "alignItems": "flex-end",
          }
        }
      >
        <ViewMoreButton
          onPress={[Function]}
        />
      </Component>
    </Component>
  </Component>
</Component>
`;

exports[`Status test suite should render properly redux action 1`] = `
<Component
  style={
    Object {
      "flex": 1,
      "width": "100%",
    }
  }
>
  <Component
    style={
      Array [
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "marginBottom": 10,
          "paddingHorizontal": 16,
        },
      ]
    }
  >
    <Title
      style={
        Array [
          Object {
            "marginRight": 6,
          },
          Object {
            "color": "#F9FAFB",
          },
        ]
      }
    >
      REDUX
    </Title>
    <Component
      style={
        Array [
          Object {
            "borderRadius": 4,
            "fontSize": 12,
            "height": 22,
            "justifyContent": "center",
            "width": 30,
          },
          Object {
            "backgroundColor": "#764ABC",
          },
          Object {
            "paddingHorizontal": 8,
          },
        ]
      }
      testID="statusCodeColor"
    >
      <Icon
        color="#F9FAFB"
        name="redux"
        size={12}
      />
    </Component>
  </Component>
  <Component
    style={
      Array [
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "marginBottom": 10,
          "paddingHorizontal": 16,
        },
      ]
    }
  >
    <Text
      style={
        Object {
          "color": "#6B7280",
        }
      }
    >
      Started at : 
    </Text>
    <Text>
      01/01/1970 - 00:00:00
    </Text>
  </Component>
  <Subheading
    style={
      Array [
        Object {
          "marginBottom": 10,
          "paddingHorizontal": 16,
          "paddingVertical": 8,
        },
        Object {
          "backgroundColor": "#374151",
          "color": "#F9FAFB",
        },
      ]
    }
  >
    Content
  </Subheading>
  <Component
    style={
      Array [
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "marginBottom": 10,
          "paddingHorizontal": 16,
        },
      ]
    }
  >
    <Text
      style={
        Object {
          "color": "#6B7280",
        }
      }
    >
      Type
    </Text>
  </Component>
  <Component
    style={
      Array [
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "marginBottom": 10,
          "paddingHorizontal": 16,
        },
      ]
    }
  >
    <Text>
      __ERROR:UNDEFINED__
    </Text>
  </Component>
  <Component
    style={
      Array [
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "marginBottom": 10,
          "paddingHorizontal": 16,
        },
      ]
    }
  >
    <Text
      style={
        Object {
          "color": "#6B7280",
        }
      }
    >
      __ERROR:UNDEFINED__ :
    </Text>
  </Component>
  <Component
    style={
      Array [
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "marginBottom": 10,
          "paddingHorizontal": 16,
        },
      ]
    }
  >
    <Component
      style={
        Array [
          Object {
            "justifyContent": "flex-start",
            "paddingHorizontal": 16,
            "paddingVertical": 8,
          },
          Object {
            "flex": 1,
            "paddingHorizontal": 0,
          },
        ]
      }
    >
      <Text>
        "This is a very long payload to be sure that we have more than 100 characters to explore all the bra...
      </Text>
      <Component
        style={
          Object {
            "alignItems": "flex-end",
          }
        }
      >
        <ViewMoreButton
          onPress={[Function]}
        />
      </Component>
    </Component>
  </Component>
</Component>
`;
